File: android-nosql/src/main/java/com/github/florent37/androidnosql/NoSqlSerializerUtils.java
package com.github.florent37.androidnosql;

import java.lang.reflect.Field;
import java.lang.reflect.ParameterizedType;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Created by florentchampigny on 29/05/2017.
 */

public class NoSqlSerializerUtils {
    public static boolean isAcceptableObject(Object object) {
        return isPrimitiveObject(object) ||
                object instanceof NoSql.Node;
    }

    public static Class genericType(Field field){
        final ParameterizedType listType = (ParameterizedType) field.getGenericType();
        final Class<?> listClass = (Class<?>) listType.getActualTypeArguments()[0];
        return listClass;
    }

    public static boolean isPrimitiveObject(Object object) {
        return object instanceof String ||
                object instanceof Integer ||
                object instanceof Double ||
                object instanceof Float ||
                object instanceof Long;
    }

    public static boolean isPrimitiveField(Field field) {
        return isPrimitive(field.getType());
    }

    public static boolean isPrimitive(Class type) {
        return
                Integer.class.isAssignableFrom(type) ||
                        Long.class.isAssignableFrom(type) ||
                        Double.class.isAssignableFrom(type) ||
                        Float.class.isAssignableFrom(type) ||
                        Boolean.class.isAssignableFrom(type) ||
                        String.class.isAssignableFrom(type);
    }

    public static boolean isCollection(Field field) {
        return List.class.isAssignableFrom(field.getType());
    }

    public static boolean isArray(Field field) {
        return field.getType().isArray();
    }

    public static boolean isInterface(Field field) {
        return field.getType().isInterface();
    }


    public static Set<Field> getAllFields(Class theClass) {
        final Set<Field> allFields = new HashSet<>();
        Class tmpClass = theClass;
        while (tmpClass != Object.class) {
            allFields.addAll(Arrays.asList(tmpClass.getDeclaredFields()));
            tmpClass = tmpClass.getSuperclass();
        }
        return allFields;
    }

}

File: android-nosql/src/main/java/com/github/florent37/androidnosql/NoSqlSerializerUtils.kt
package com.github.florent37.androidnosql

import java.lang.reflect.Field
import java.lang.reflect.ParameterizedType
import java.util.*

fun Any?.isValueObject(): Boolean {
    return if(this == null)
        false
    else this.isPrimitiveObject() || this is NosqlElement.Node
}

fun Field.genericType(): Class<*> {
    val listType = this.genericType as ParameterizedType
    return listType.actualTypeArguments[0] as Class<*>
}

fun Any?.isPrimitiveObject(): Boolean {
    return this is String ||
            this is Int ||
            this is Double ||
            this is Float ||
            this is Long
}

fun Field.isPrimitiveField(): Boolean {
    return this.type.isPrimitiveType()
}

fun Class<*>?.isPrimitiveType(): Boolean {
    return Int::class.java.isAssignableFrom(this) ||
            Long::class.java.isAssignableFrom(this) ||
            Double::class.java.isAssignableFrom(this) ||
            Float::class.java.isAssignableFrom(this) ||
            Boolean::class.java.isAssignableFrom(this) ||
            String::class.java.isAssignableFrom(this)
}

fun Field.isCollection(): Boolean {
    return MutableList::class.java.isAssignableFrom(this.type)
}

fun Field.isArray(): Boolean {
    return this.type.isArray
}

fun Field.isInterface(): Boolean {
    return this.type.isInterface
}

fun Class<*>?.getAllFields(): Set<Field> {
    val allFields: MutableSet<Field> = HashSet()

    this?.let { initialClass ->
        var tmpClass = initialClass
        while (tmpClass != Any::class.java) {
            allFields.addAll(listOf(*tmpClass.declaredFields))
            tmpClass = tmpClass.superclass
        }
    }

    return allFields
}
